File: baekjoon_all/02000+/boj_2580.java
// Solve 2024-03-28

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_2580 {

    static int[][] board = new int[9][9];
    static boolean[][] fixed = new boolean[9][9];
    static boolean[][] rowCheck = new boolean[9][10];
    static boolean[][] colCheck = new boolean[9][10];
    static boolean[][][] squareCheck = new boolean[3][3][10];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        for (int x = 0; x < 9; x++) {
            st = new StringTokenizer(br.readLine(), " ");

            for (int y = 0; y < 9; y++) {
                int curNum = st.nextToken().charAt(0) - '0';

                if (curNum > 0) {
                    board[x][y] = curNum;
                    fixed[x][y] = true;
                    rowCheck[x][curNum] = true;
                    colCheck[y][curNum] = true;
                    squareCheck[x / 3][y / 3][curNum] = true;
                }
            }
        }

        recur(0);

        StringBuilder sb = new StringBuilder();

        for (int x = 0; x < 9; x++) {
            for (int y = 0; y < 9; y++) {
                sb.append(board[x][y]).append(" ");
            }

            sb.append("\n");
        }

        System.out.print(sb);
        br.close();
    }

    static boolean recur(int depth) {
        if (depth == 81) {
            return true;
        }

        int x = depth / 9;
        int y = depth % 9;

        if (fixed[x][y]) {
            if (recur(depth + 1)) {
                return true;
            }

            return false;
        }

        for (int curNum = 1; curNum <= 9; curNum++) {
            if (rowCheck[x][curNum] || colCheck[y][curNum] || squareCheck[x / 3][y / 3][curNum]) {
                continue;
            }

            rowCheck[x][curNum] = true;
            colCheck[y][curNum] = true;
            squareCheck[x / 3][y / 3][curNum] = true;
            board[x][y] = curNum;

            if (recur(depth + 1)) {
                return true;
            }

            rowCheck[x][curNum] = false;
            colCheck[y][curNum] = false;
            squareCheck[x / 3][y / 3][curNum] = false;
        }

        return false;
    }

}

File: baekjoon_all/02000+/boj_2580.cpp
// Solve 2022-08-01
// Update 2024-03-28

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int board[9][9];
bool board_fixed[9][9];
bool row_check[9][10];
bool col_check[9][10];
bool square_check[3][3][10];

bool recur(int depth = 0) {
    if (depth == 81) return true;

    int x = depth / 9;
    int y = depth % 9;

    if (board_fixed[x][y]) {
        if (recur(depth + 1)) {
            return true;
        }

        return false;
    }

    for (int cur_num = 1; cur_num <= 9; cur_num++) {
        if (row_check[x][cur_num] || col_check[y][cur_num] || square_check[x / 3][y / 3][cur_num]) {
            continue;
        }

        row_check[x][cur_num] = true;
        col_check[y][cur_num] = true;
        square_check[x / 3][y / 3][cur_num] = true;
        board[x][y] = cur_num;

        if (recur(depth + 1)) {
            return true;
        }

        row_check[x][cur_num] = false;
        col_check[y][cur_num] = false;
        square_check[x / 3][y / 3][cur_num] = false;
    }

    return false;
}

int main() {
    FASTIO;

    for (int x = 0; x < 9; x++) {
        for (int y = 0; y < 9; y++) {
            cin >> board[x][y];

            if (board[x][y] > 0) {
                board_fixed[x][y] = true;
                row_check[x][board[x][y]] = true;
                col_check[y][board[x][y]] = true;
                square_check[x / 3][y / 3][board[x][y]] = true;
            }
        }
    }

    recur();

    for (int x = 0; x < 9; x++) {
        for (int y = 0; y < 9; y++) {
            cout << board[x][y] << ' ';
        }

        cout << '\n';
    }

    return 0;
}
